### ISSUE TITLE
REST Controller

### ISSUE THREAD

**Issue by neo22s** [koseven/koseven#284]
I've developed this API_Controller some years ago:
https://github.com/yclas/yclas/blob/master/oc/classes/api/controller.php

Some controller usage examples:
https://github.com/yclas/yclas/tree/master/oc/classes/controller/api

Authentication: In our case we create a token per user:
https://github.com/yclas/yclas/blob/master/oc/classes/auth/oc.php#L188:

Route that could be added as default:
https://github.com/yclas/yclas/blob/master/oc/config/routes.php#L264

We will need to review all code if works out of the box and make a good proper documentation with tons of examples.

Not sure if this is something we should have out of the box or in a module, but a lightweight framework as Koseven should be used by many as backend for their apps and currently it is not easy to do it and I believe it's due to the lack of such a "basic" thing ;)

**Comment by Sebjugate** [comment]
Cool. I agree that this is something I suspect a lot of apps could use or have already implemented in their own way. I have implemented something similar for my own apps.

If it is to be included, I would favor a module.

**Comment by toitzi** [comment]
Look at my extension: https://github.com/toitzi/koseven-rest

It's pretty much that..isn't it?

**Comment by toitzi** [comment]
@neo22s I agree. REST functions is a must have for me...especially when creating APIs

I will update the docs.

**Comment by toitzi** [comment]
@neo22s ok, should i do a PR?

**Comment by neo22s** [comment]
can you add tests please?

**Comment by toitzi** [comment]
yes

**Comment by piotrbaczek** [comment]
I can look into this this weekend @toitzi if you don't have time.
Let's create Controller_Api_Base that will extend from Controller

**Comment by toitzi** [comment]
@piotrbaczek would be great thanks!

**Comment by toitzi** [comment]
@WinterSilence @piotrbaczek @neo22s  I created a PR (#374) regarding this (not final, just a quick implementation),
i took some parts of your code @WinterSilence and some of the original one and put it together, i also added a REST_Response class which allows passing arrays to the body (very common for REST Api's)....please let me think if you think this is the right direction...here are 2 examples of how to use it:

### For both examples
Simply change inside `Controller_Welcome`:
`extends Controller` to `extends Controller_REST`.

### 1
That's already it!

Now the output should be:
`{"body":"hello, world!"}` (with `application/json` as header)

### 2
You can also change the line `$this->response->body('hello, world!');` to something like:
`$this->response->body(['id' => 1, 'name' => 'Koseven']);`

Now the output should be:
`{"id":1,"name":"Koseven"}` (with `application/json` as header)

### Conclusion
Pretty easy isn't it?
_Please keep in mind that this was just a quick pr for how we could implement it, configuration, documentation, different formats need to be added...._

I would really appriciate if you could give small feedback if that's the way we should implement it.

@WinterSilence in your previous comment you said:

> _format_* methods need move in adapter-classes

Can you give me a quick example?

Thanks!

**Comment by WinterSilence** [comment]
@toitzi i use callbacks for this, see `*_format` properties in my version

**Comment by toitzi** [comment]
Just needs Unittests, there is other stuff to do and in favour of releasing 4.0 i am gonna do them later
